returnn/flow.py

```diff
@@ -52,66 +88,114 @@ def make_precomputed_hybrid_tf_feature_flow(
+    rasr_config.output_map.info_0.param_name = "log-posteriors"
+    rasr_config.output_map.info_0.tensor_name = f"{output_layer_name}/output_batch_major"
 
-    tf_flow.config[tf_fwd].loader.type = "meta"
-    tf_flow.config[tf_fwd].loader.meta_graph_file = tf_graph
-    tf_flow.config[tf_fwd].loader.saved_model_file = tf_checkpoint
+    rasr_config.loader.type = "meta"
+    rasr_config.loader.meta_graph_file = tf_graph
+    rasr_config.loader.saved_model_file = tf_checkpoint
     if native_ops is not None:
         if isinstance(native_ops, list):
-            tf_flow.config[tf_fwd].loader.required_libraries = DelayedJoin(native_ops, ";")
+            rasr_config.loader.required_libraries = DelayedJoin(native_ops, ";")
         else:
-            tf_flow.config[tf_fwd].loader.required_libraries = native_ops
+            rasr_config.loader.required_libraries = native_ops
+    return make_precomputed_hybrid_feature_flow(
+        backend="tf",
+        rasr_config=rasr_config,
+        fwd_input_name=tf_fwd_input_name,
+    )
 
-    return tf_flow
 
+def make_precomputed_hybrid_onnx_feature_flow(
+    onnx_model: tk.Path,
+    io_map: Dict[str, str],
+    onnx_fwd_input_name: str = "onnx-fwd-input",
+    cpu: int = 1,
+) -> rasr.FlowNetwork:
+    """
+    Create the feature flow for a simple ONNX network that predicts frame-wise outputs,
+    see make_precomputed_hybrid_feature_flow.
 
-def add_tf_flow_to_base_flow(
+    With the config settings:
+
+        [flf-lattice-tool.network.recognizer.feature-extraction.onnx-fwd.io-map]
+        features      = data
+        features-size = data_len
+        output        = classes
+
+        [flf-lattice-tool.network.recognizer.feature-extraction.onnx-fwd.session]
+        file                 = <onnx_file>
+        inter-op-num-threads = 2
+        intra-op-num-threads = 2
+
+    :param onnx_model: usually the output of a OnnxExportJob
+    :param io_map: e.g. {"features": "data", "output": "classes"}
+    :param onnx_fwd_input_name: naming for the onnx network input, usually no need to be changed
+    :param cpu: number of CPUs to use
+    :return: onnx-forward node flow with output link and related config
+    """
+
+    rasr_config = rasr.RasrConfig()
+    rasr_post_config = rasr.RasrConfig()
+    for k, v in io_map.items():
+        rasr_config.io_map[k] = v
+
+    rasr_config.session.file = onnx_model
+    rasr_post_config.session.inter_op_num_threads = cpu
+    rasr_post_config.session.intra_op_num_threads = cpu
+
+    return make_precomputed_hybrid_feature_flow(
+        backend="onnx",
+        rasr_config=rasr_config,
+        rasr_post_config=rasr_post_config,
+        fwd_input_name=onnx_fwd_input_name,
+    )
+
+
+def add_fwd_flow_to_base_flow(
     base_flow: rasr.FlowNetwork,
-    tf_flow: rasr.FlowNetwork,
-    tf_fwd_input_name: str = "tf-fwd-input",
+    fwd_flow: rasr.FlowNetwork,
+    fwd_input_name: str = "fwd-input",
```

Comment:
The default name here is `"fwd-input"`. 
The default names for the `make_precomputed_hybrid_tf_feature_flow` and `make_precomputed_hybrid_onnx_feature_flow` are `"tf-fwd-input"` and `"onnx-fwd-input"` respectively. 

These names always overwrite the default of `make_precomputed_hybrid_feature_flow` (i.e. `"fwd-input"`) and I don't believe that one can use `make_precomputed_hybrid_feature_flow` standalone, because the config will be missing.

I conclude, that the default naming parameters will never match. Or am I missing something?

As the name will anyway never be seen and is arbitrary, I would propose to make your lives easier and just call it `"fwd-input"` in all the cases (except for the legacy `add_tf_flow_to_base_flow`)

Reply:
Yes, for onnx we can just name it `"fwd-input". For `add_tf_flow_to_base_flow` we should keep the old name in order not to break the hashes. That's what you propose, right?

Reply:
mh.. would there be a way to auto-detect it? e.g. 
```
input_name = list(fwd_flow.inputs)[0]
```
(I think the flow has just a single input)

Reply:
I'm not sure if this is always guaranteed. In the way it is now, the user also never needs to set it. Do you still prefer auto-detecting it?

Reply:
>  In the way it is now, the user also never needs to set it.

if they use the new  `add_fwd_flow_to_base_flow` with TF, then they still have to set it, right?

> I'm not sure if this is always guaranteed. Do you still prefer auto-detecting it?

I do not have a strong opinion and would let that decision to the people that will use this functions.

Reply:
> > In the way it is now, the user also never needs to set it.
>
> if they use the new add_fwd_flow_to_base_flow with TF, then they still have to set it, right?

Depends. First, this function is used to avoid code duplications between tf and onnx. Using it independently of either of them is currently no use case. Second, if (as discussed in the other comment) we make `"fwd-input"` the default for everything that isn't tf, the user can always stick to the default here.

Reply:
@michelwi do you want changes here? Otherwise we could merge.

Reply:
Thanks!